[AMDGPU] Enable overwrite ALU bit in sched.barrier mask

jplehr opened the pull request:
The sched.barrier takes a bit mask that determines which instruction categories are allowed to cross the inserted sched.barrier during the igrouplp scheduling pass.

Currently, a set ALU bit results in allowing all ALU instructions to move across the barrier, independent of whether more specific bits have been specified.
The documentation is silent about the semantics in that case.

This PR changes the current handling: When a mask contains both a set ALU bit and a set more-specific bit, the more specific bit is respected and the ALU bit does *not imply* all other bits.

Current:
0x00000005 -- 0101 set ALU and SALU bit. Currently the ALU bit implies
                   SALU and VALU and MFMA to be set.

New:
0x00000005 -- 0101 set ALU and SALU bit. SALU bit set, therefore ALU bit
                   is ignored and only SALU bit is considered.

jplehr committed:
[AMDGPU] Enable overwrite ALU bit in sched.barrier mask

The sched.barrier takes a bit mask that determines which instruction
categories are allowed to cross the inserted sched.barrier during the
igrouplp scheduling pass.

Currently, a set ALU bit results in allowing all ALU instructions to
move across the barrier, independent of whether more specific bits
have been specified.
The documentation is silent about the semantics in that case.

This PR changes the current handling: When a mask contains both a set
ALU bit and a set more-specific bit, the more specific bit is respected
and the ALU bit does *not imply* all other bits.

Current:
0x00000005 -- 0101 set ALU and SALU bit. Currently the ALU bit implies
                   SALU and VALU and MFMA to be set.

New:
0x00000005 -- 0101 set ALU and SALU bit. SALU bit set, therefore ALU bit
                   is ignored and only SALU bit is considered.

jplehr commented:
I am aware that this needs a/several test(s) and I would mostly like to get some feedback on whether this is an acceptable approach.

hidekisaito reviewed (approved):
LGTM

jrbyrnes commented:
I think that if we want to do this, we need to be consistent across all different masks. For example, why do we constrain for SALU if our mask has both SALU and ALU, but we don't constrain for VMEM_READ if we have both VMEM_READ and VMEM.

Also, I'm not sure what the intended use case is here.

Needs tests.

jplehr commented:
> I think that if we want to do this, we need to be consistent across all different masks. For example, why do we constrain for SALU if our mask has both SALU and ALU, but we don't constrain for VMEM_READ if we have both VMEM_READ and VMEM.

Excellent points. I agree that it should be handled consistently. If you are ok with the direction that this change is going, then I can go ahead and do the changes and clarifications in the documents.

> 
> Also, I'm not sure what the intended use case is here.

The intended use is coming from Triton where I think they want to set everything and then selectively unset certain bits, e.g., MFMA.

I think the overall expectation is: If you specify "everything" (ALU bit) but then want to selectively unset something, the latter should take precedence.

arsenm reviewed (changes requested):
Needs tests

jplehr committed:
Add MEM handling

jplehr committed:
Adds test

Various different sched.barrier masks are tested for their inverse.
The inverse is used to create a sched.group in which the instructions
that are not allowed to move are contained.

jplehr committed:
Edit documentation

jplehr committed:
Clean tests a bit

jplehr committed:
Fix formatting issue

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:
```diff
@@ -2630,8 +2636,15 @@ IGroupLPDAGMutation::invertSchedBarrierMask(SchedGroupMask Mask) const {
     InvertedMask &= ~SchedGroupMask::ALU;
 
+  // When given specific bits overrule the more general MEM type.
+  bool HasConcreteMemClassSpecified =
```
The concrete VMEM should be distinct from concrete DS

jrbyrnes reviewed (commented):
> Currently, a set ALU bit results in allowing all ALU instructions to move across the barrier, independent of whether more specific bits have been specified. The documentation is silent about the semantics in that case.
> 
> Current: 0x00000005 -- 0101 set ALU and SALU bit. Currently the ALU bit implies SALU and VALU and MFMA to be set.

The current AMDGPUUsage document states: "The parameter is a mask for the instruction types that can cross the intrinsic." To me, this implies that the mask of 0x00000005 will allow ALU and SALU instructions to cross the barrier -- which means VALU, MFMA, SALU -- the SALU bit is redundant.

If the user wants only the SALU to cross the barrier, can't they just set the SALU bit?

I think there is some weirdness that occurs when the user tries to invert the mask to block instructions from passing, but I think I need some more context to understand what problem this is addressing. What is the user trying to do?

jplehr commented:
Closing this in favor of new NFC documentation update.
